### Importing the Datasets 

In [1]:
import numpy as np  
import pandas as pd 
import math

In [2]:
df1 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2001_to_2004.csv", error_bad_lines = False)

b'Skipping line 1513591: expected 23 fields, saw 24\n'
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (17,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df2 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2005_to_2007.csv", error_bad_lines = False)

b'Skipping line 533719: expected 23 fields, saw 24\n'


In [4]:
df3 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2008_to_2011.csv", error_bad_lines = False)

b'Skipping line 1149094: expected 23 fields, saw 41\n'


In [9]:
df4 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2012_to_2017.csv", error_bad_lines = False)

### Analyzing Missing Data 

In [10]:
'''Takes in a dataframe and returns a dataframe with percentages of missing data of each column
'''
def missingDataSummary(df):
    dfInitial = pd.DataFrame(index = [df.name], columns =df.columns)
    for column in df.columns:
        #print("Column name: " + column )
        numNas = df[column].isnull().sum()
        numObservations = len(df[column])
        missingDataPercentage = numNas/numObservations *100
        dfInitial[column][0] = missingDataPercentage
        #print(str(missingDataPercentage) + "% missing") 
    return dfInitial 

In [11]:
#set the names of the dataframes for convenience 
df1.name = 'df1'
df2.name = 'df2'
df3.name = 'df3'
df4.name = 'df4'

In [12]:
#retrieve info on missing data for each data frame 
df1_missingSummary = missingDataSummary(df1)
df2_missingSummary  = missingDataSummary(df2)
df3_missingSummary = missingDataSummary(df3)
df4_missingSummary = missingDataSummary(df4)

#concatenate missingData summary info into one
df_missingSummaries = pd.concat([df1_missingSummary, df2_missingSummary, df3_missingSummary, df4_missingSummary], axis = 0)

#display all the missing data summary
pd.set_option('display.max_columns', None)
df_missingSummaries

Unnamed: 0 ID  Case Number Date Block IUCR Primary Type Description  \
df1          0  0            0    0     0    0            0           0   
df2          0  0            0    0     0    0            0           0   
df3          0  0  0.000223155    0     0    0            0           0   
df4          0  0  6.86477e-05    0     0    0            0           0   

    Location Description Arrest Domestic Beat     District         Ward  \
df1          0.000831811      0        0    0  0.000103976      36.3986   
df2           0.00133523      0        0    0  0.000267045  0.000801135   
df3             0.010823      0        0    0   0.00308698   0.00234313   
df4             0.113818      0        0    0  6.86477e-05  0.000961067   

    Community Area FBI Code X Coordinate Y Coordinate Year Updated On  \
df1        36.4046        0      1.59557      1.59557    0          0   
df2      0.0186397        0     0.488906     0.488906    0          0   
df3      0.0541152        0      1.06538      1.06538    0          0   
df4     0.00274591        0      2.54566      2.54566    0          0   

     Latitude Longitude  Location  
df1   1.59557   1.59562   1.59562  
df2  0.488906  0.488906  0.488906  
df3   1.06538   1.06538   1.06538  
df4   2.54566   2.54566   2.54566

### Concatenating Four Datasets --> dfTotal dataset 

In [13]:
#concatenate 4 data sets to make one total data set form years range 2001-2017
dfTotal = pd.concat([df1, df2, df3, df4], axis = 0)

In [14]:
#checking if the concatenation indeed happened
print("Are the number of columns maintained?")
print(len(df1.columns) == len(df2.columns) ==  len(df3.columns) == len(df4.columns) == len(dfTotal.columns))

print("Are the number of observations of all data sets equal to the dfTotal")
print(len(df1['ID']) + len(df2['ID']) + len(df3['ID']) + len(df4['ID']) == len(dfTotal['ID']))

dfTotal.columns

Are the number of columns maintained?
True
Are the number of observations of all data sets equal to the dfTotal
True


Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [15]:
dfTotal.head()

Unnamed: 0       ID Case Number                    Date  \
0         879  4786321    HM399414  01/01/2004 12:01:00 AM   
1        2544  4676906    HM278933  03/01/2003 12:00:00 AM   
2        2919  4789749    HM402220  06/20/2004 11:00:00 AM   
3        2927  4789765    HM402058  12/30/2004 08:00:00 PM   
4        3302  4677901    HM275615  05/01/2003 01:00:00 AM   

                 Block  IUCR                Primary Type  \
0    082XX S COLES AVE  0840                       THEFT   
1      004XX W 42ND PL  2825               OTHER OFFENSE   
2  025XX N KIMBALL AVE  1752  OFFENSE INVOLVING CHILDREN   
3   045XX W MONTANA ST  0840                       THEFT   
4   111XX S NORMAL AVE  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   Beat  District  Ward  Community Area FBI Code  X Coordinate Y Coordinate  \
0   424       4.0   7.0            46.0       06           NaN          NaN   
1   935       9.0  11.0            61.0       26     1173974.0  1.87676e+06   
2  1413      14.0  35.0            22.0       20           NaN          NaN   
3  2521      25.0  31.0            20.0       06           NaN          NaN   
4  2233      22.0  34.0            49.0       06     1174948.0  1.83105e+06   

     Year              Updated On Latitude  Longitude  \
0  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
1  2003.0  04/15/2016 08:55:02 AM  41.8172 -87.637328   
2  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
3  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
4  2003.0  04/15/2016 08:55:02 AM  41.6918 -87.635116   

                        Location  
0                            NaN  
1  (41.817229156, -87.637328162)  
2                            NaN  
3                            NaN  
4  (41.691784636, -87.635115968)

### Conversion of Time Data --> Time Series for easy sorting in pandas df 

In [16]:
df1.sort_values(by='Year')

Unnamed: 0       ID Case Number                    Date  \
1602847     3629582  1423259     G139165  03/10/2001 11:30:00 PM   
386628      3926953  1797222     G622842  10/16/2001 04:00:00 PM   
379441      3919766  1788231     G603480  10/08/2001 09:15:00 AM   
379440      3919765  1788230     G588035  10/01/2001 11:45:00 AM   
379439      3919764  1788228     G604692  10/08/2001 07:00:10 PM   
379438      3919763  1788227     G608421  10/10/2001 12:44:27 PM   
379437      3919762  1788225     G607431  10/09/2001 10:20:28 PM   
379436      3919761  1788224     G588034  10/01/2001 11:43:00 AM   
379435      3919760  1788222     G608283  10/10/2001 11:41:25 AM   
379434      3919759  1788221     G588513  10/01/2001 02:10:00 PM   
379433      3919758  1788218     G610767  10/05/2001 07:00:00 AM   
379432      3919757  1788217     G603979  10/08/2001 01:30:00 PM   
379431      3919756  1788216     G610432  10/10/2001 02:00:00 PM   
379430      3919755  1788215     G608337  08/24/2001 09:00:00 AM   
379429      3919754  1788214     G588033  10/01/2001 10:30:00 AM   
379442      3919767  1788234     G609911  10/11/2001 12:15:00 AM   
379428      3919753  1788213     G604683  10/08/2001 06:55:11 PM   
379426      3919751  1788208     G588032  10/01/2001 11:40:00 AM   
379425      3919750  1788207     G603942  10/08/2001 01:05:00 PM   
379424      3919749  1788205     G588610  10/01/2001 03:00:00 PM   
379423      3919748  1788204     G605259  10/08/2001 11:40:00 PM   
379422      3919747  1788203     G604681  10/08/2001 06:04:36 PM   
379421      3919746  1788202     G603023  10/08/2001 12:15:00 AM   
379420      3919745  1788201     G607810  10/10/2001 06:35:18 AM   
379419      3919744  1788200     G610801  10/09/2001 06:00:00 PM   
379418      3919743  1788199     G605412  10/09/2001 03:58:12 AM   
379417      3919742  1788198     G588031  10/01/2001 11:50:00 AM   
379416      3919741  1788197     G608137  10/10/2001 10:30:00 AM   
379415      3919740  1788196     G600377  10/06/2001 04:15:00 PM   
379414      3919739  1788193     G588142  10/01/2001 11:15:00 AM   
...             ...      ...         ...                     ...   
1663497     5263662  3394796    HK442661  06/20/2004 01:40:00 AM   
1663496     5263661  3394794    HK447967  06/21/2004 12:00:00 AM   
1663495     5263660  3394793    HK447633  06/20/2004 12:30:00 PM   
1663494     5263659  3394792    HK449598  06/22/2004 10:30:00 AM   
1663503     5263668  3394803    HK448110  06/22/2004 02:00:00 PM   
1663540     5263705  3394844    HK449214  06/23/2004 01:15:00 AM   
1663515     5263680  3394816    HK441894  06/05/2004 07:30:00 PM   
1663517     5263682  3394818    HK449681  06/22/2004 11:00:00 PM   
1663538     5263703  3394842    HK449848  06/23/2004 11:10:00 AM   
1663537     5263702  3394841    HK444410  06/21/2004 12:20:00 AM   
1663536     5263701  3394840    HK439158  06/18/2004 12:51:02 PM   
1663535     5263700  3394839    HK446865  06/22/2004 12:20:00 AM   
1663533     5263698  3394836    HK442733  06/20/2004 02:50:00 AM   
1663532     5263697  3394835    HK446816  06/21/2004 11:45:00 PM   
1663531     5263696  3394834    HK439925  06/16/2004 08:30:00 PM   
1663530     5263695  3394833    HK446733  06/21/2004 05:10:00 PM   
1663529     5263694  3394832    HK448354  06/22/2004 04:55:00 PM   
1663516     5263681  3394817    HK446542  06/21/2004 08:50:00 PM   
1663528     5263693  3394831    HK412919  06/06/2004 04:28:23 PM   
1663526     5263691  3394829    HK446718  06/21/2004 10:19:41 PM   
1663525     5263690  3394828    HK436733  06/17/2004 11:08:52 AM   
1663524     5263689  3394827    HK412675  06/06/2004 02:25:00 PM   
1663523     5263688  3394826    HK442898  06/20/2004 06:16:00 AM   
1663522     5263687  3394825    HK448481  06/22/2004 06:00:00 PM   
1663521     5263686  3394823    HK439498  06/18/2004 03:13:28 PM   
1663520     5263685  3394821    HK442803  06/20/2004 04:10:00 AM   
1663519     5263684  3394820    HK412603  06/06/2004 01:45:00 PM 

In [17]:
def changeToDateTime(df, columnName, timeFormat): 
    df[columnName] = pd.to_datetime(df[columnName], format = timeFormat)

In [18]:
dateFormat = '%m/%d/%Y %I:%M:%S %p' 
yearFormat = '%Y.0'

In [19]:
changeToDateTime(df1, 'Date', dateFormat)

In [62]:
#changeToDateTime(df1, 'Year', yearFormat)

### Exploration of null values and time dataset 

In [92]:
df2003 = df1.loc[df1['Date'].dt.year == 2003] #allows easy slicing of data by time

In [110]:
import math

In [20]:
df1WardNull = df1.loc[df1['Ward'].isnull()] #choose all null values

In [133]:
len(df1WardNull)

700132

In [21]:
df1WardNull

Unnamed: 0       ID Case Number                Date  \
1332         808956  1954658    HH143979 2002-01-23 17:30:00   
1450         814436      636     G008548 2001-01-04 22:30:00   
1457         814500      637     G005960 2001-01-06 10:35:00   
1468         814582      638     G008793 2001-01-05 16:22:00   
1507         815017      642     G013102 2001-01-07 00:48:00   
1516         815109      643     D362053 2001-01-04 22:05:00   
1537         815300      645     G019063 2001-01-09 21:34:00   
1565         815711      649     G022582 2001-01-11 14:25:00   
1570         815792      650     G023235 2001-01-11 18:46:00   
1580         815881      651     G023235 2001-01-11 21:10:00   
1618         816387      657     G030711 2001-01-15 09:59:00   
1632         816617      660     G032650 2001-01-16 09:00:00   
1658         817257      665     G042741 2001-01-21 01:09:00   
1687         817577      669     G052279 2001-01-25 19:38:00   
1746         818345      674     G042867 2001-01-30 02:10:00   
1784         818781      678     G069278 2001-02-03 00:53:00   
1905         820006      693     G103123 2001-02-20 11:10:00   
1908         820061      694     G108432 2001-02-23 12:19:00   
1918         820246      696     G109745 2001-02-24 03:55:00   
1927         820337      697     G110056 2001-02-24 11:05:00   
1933         820406      698     G107568 2001-02-22 16:45:00   
1954         820583      701     G118039 2001-02-28 17:36:00   
2034         821714      712     G132945 2001-03-08 01:00:00   
2084         822279      718     G139050 2001-03-10 21:16:00   
2103         822460      721     G143991 2001-03-13 11:00:00   
2124         822651      724     G159652 2001-03-20 18:17:00   
2137         822773      727     G164500 2001-03-23 11:35:00   
2204         823337      738     G179864 2001-03-30 09:22:00   
2234         823508      742     G187460 2001-04-02 21:45:00   
2238         823564      743     G188820 2001-04-03 15:05:00   
...             ...      ...         ...                 ...   
1602834     3629569  1423240     G141190 2001-03-12 01:30:00   
1602835     3629570  1423241     G141080 2001-03-11 13:00:00   
1602836     3629571  1423242     G145333 2001-03-13 21:40:00   
1602837     3629572  1423243     G128245 2001-03-05 18:11:38   
1602838     3629573  1423245     G144065 2001-03-13 11:17:31   
1602839     3629574  1423246     G140315 2001-03-11 15:10:00   
1602840     3629575  1423247     G143699 2001-03-09 13:30:00   
1602841     3629576  1423248     G143637 2001-03-13 07:00:00   
1602842     3629577  1423249     G142018 2001-03-12 11:40:00   
1602843     3629578  1423251     G127313 2001-03-05 10:18:34   
1602844     3629579  1423252     G144685 2001-03-13 08:10:00   
1602845     3629580  1423257     G128305 2001-03-05 17:54:32   
1602846     3629581  1423258     G135926 2001-03-09 11:55:00   
1602847     3629582  1423259     G139165 2001-03-10 23:30:00   
1839699     5440832  3728914    HK833343 2004-12-29 01:28:00   
1901401     5503150  3720895    HK823653 2004-12-18 23:00:00   
1909558     5513657  3736725    HL103716 2004-12-29 14:00:00   
1923429     6196323  1331522     G028902 2001-01-09 14:00:00   
1923430     6196324  1374353     G080293 2001-02-08 12:00:00   
1923431     6196325  1422013     G143930 2001-03-13 10:00:00   
1923432     6196326  1476586     G209906 2001-04-12 23:30:00   
1923433     6196327  1537467     G289974 2001-05-19 23:15:00   
1923434     6196328  1631874     G413687 2001-07-15 04:30:00   
1923435     6196329  1639542     G413681 2001-07-15 04:00:00   
1923436     6196330  1741866     G547008 2001-09-01 16:00:00   
1923437     6196331  1833095     G668623 2001-11-06 11:20:00   
1923438     6196332  1839832     G676174 2001-11-09 14:30:00   
1923439     6196333  1899039     G752382 2001-12-16 19:20:00   
1923440     6196334  1908900     G767033 2001-12-23 21:30:00   
1923455     6213128  1418407     G139151 2001-03-10 20:50:00   

                                  Block 

In [22]:
ward2001Nan = df1.loc[(df1['Ward'].isnull()) &( df1['Date'].dt.year == 2001)]

In [23]:
len(ward2001Nan)

563443

In [24]:
df1['Ward'].isnull().sum()

700132

In [25]:
df2001 = len(df1.loc[df1['Date'].dt.year == 2001])

In [93]:
len(df2003)

475913

In [94]:
df2003

Unnamed: 0        ID Case Number                Date  \
1              2544   4676906    HM278933 2003-03-01 00:00:00   
4              3302   4677901    HM275615 2003-05-01 01:00:00   
7              4502   4679521    HM216293 2003-03-15 00:00:00   
9              4904   4680124    HM282389 2003-01-01 00:00:00   
10             6333   4682244    HM282286 2003-06-05 00:00:00   
12             7992   4684808    HM286351 2003-01-01 00:00:00   
14             9317   4686904    HM288497 2003-11-01 08:00:00   
15             9585   4687321    HM287080 2003-11-15 09:00:00   
16             9992   4688004    HM290677 2003-01-02 08:00:00   
22            15379   4695836    HM294072 2003-10-31 14:00:00   
23            15596   4696148    HM300702 2003-01-01 09:00:00   
24            15791   4696401    HM301419 2003-01-01 00:00:00   
26            17867   4699493    HM302756 2003-01-01 12:00:00   
31            22131   4706080    HM306283 2003-12-10 09:00:00   
34            23659   4708622    HM312347 2003-05-23 09:00:00   
35            23921   4709062    HM313501 2003-06-18 09:00:00   
36            24196   4709502    HM314530 2003-12-01 00:01:00   
37            24886   4710526    HM274261 2003-08-01 09:00:00   
39            25378   4711465    HM315884 2003-07-07 00:00:00   
40            25570   4711723    HM317057 2003-08-09 12:00:00   
47            30616   4718225    HM319881 2003-11-01 12:35:00   
52            33684   4722424    HM325827 2003-10-01 00:00:00   
53            34838   4724079    HM328405 2003-08-03 00:01:00   
54            35022   4724354    HM328186 2003-07-01 00:01:00   
62            40224   4731694    HM337822 2003-02-15 00:01:00   
64            41462   4733593    HM342098 2003-07-30 00:00:00   
65            41617   4733828    HM342754 2003-09-01 09:00:00   
66            42804   4735661    HM343982 2003-07-01 00:00:00   
69            43844   4737355    HM341887 2003-02-01 12:00:00   
70            44045   4737684    HM343580 2003-06-01 09:00:00   
...             ...       ...         ...                 ...   
1923396     6100708   2862812    HJ520923 2003-07-26 21:30:00   
1923397     6101333  10617254    HZ368306 2003-06-03 09:00:00   
1923399     6113013  10635716    HZ388106 2003-08-29 00:00:00   
1923401     6122464  10632486    HZ384585 2003-01-01 08:00:00   
1923406     6145457   2973029    HJ666435 2003-10-02 00:30:00   
1923407     6151148   3062540    HJ774578 2003-11-22 01:55:36   
1923413     6161979   2863910    HJ527386 2003-07-29 18:44:59   
1923414     6162556  10714557    HZ472298 2003-01-02 08:00:00   
1923417     6173185  10726479    HZ485656 2003-01-01 08:00:00   
1923419     6173330  10729676    HZ488422 2003-01-01 00:00:00   
1923420     6177485      2335    HJ572458 2003-08-19 16:30:00   
1923421     6178265  10557048    HZ304217 2003-02-04 00:00:00   
1923446     6196345   2613260    HJ212102 2003-03-01 12:15:00   
1923447     6196346   3056940    HJ774136 2003-11-21 22:00:00   
1923451     6199595  10765725    HZ530828 2003-10-01 09:00:00   
1923454     6208908   2608267    HJ201952 2003-02-24 10:41:04   
1923458     6217888  10793112    HZ562371 2003-02-18 09:00:00   
1923461     6228059  10813751    JA113073 2003-03-17 20:00:00   
1923464     6236972   4648680    HM245882 2003-11-07 14:00:00   
1923465     6237453   4649433    HM245788 2003-06-01 00:01:00   
1923472     6240452   4653544    HM252263 2003-12-01 14:00:00   
1923479     6242249   4656149    HM255782 2003-09-01 12:00:00   
1923480     6242644   4656654    HM254485 2003-07-01 23:00:00   
1923481     6243001   4657239    HM257155 2003-01-01 00:00:00   
1923483     6243432   4657991    HM257675 2003-01-01 08:00:00   
1923490     6245589   4661263    HM247797 2003-04-01 20:00:00   
1923501     6249176   4666227    HM259200 2003-08-01 09:00:00   
1923506     6250651   4668393    HM270006 2003-02-01 09:00:00   
1923507     6252118   4670514    HM268786 2003-06-15 21:00:00   
1923511     6252606   4671197    HM270817 20

In [26]:
len(df1)

1923515

### Exploartion of Unique Observations/Categories 

In [29]:
'''Takes in a dataframe and returns a dataframe with the corresponding
unique observations made in each column
'''
def uniqueObservations(df):
    dfInitial = pd.DataFrame(index = [0], columns =df.columns)
    for column in df.columns:
        uniqueCounts = df[column].unique()
        numObservations = len(df[column])
        dfInitial[column][0] = uniqueCounts
    return dfInitial 

In [30]:
dfTotal.name = 'dfTotal'

In [31]:
dftotal_unique = uniqueObservations(dfTotal)

In [32]:
dftotal_unique

Unnamed: 0  \
0  [879, 2544, 2919, 2927, 3302, 3633, 3756, 4502...   

                                                  ID  \
0  [4786321, 4676906, 4789749, 4789765, 4677901, ...   

                                         Case Number  \
0  [HM399414, HM278933, HM402220, HM402058, HM275...   

                                                Date  \
0  [01/01/2004 12:01:00 AM, 03/01/2003 12:00:00 A...   

                                               Block  \
0  [082XX S COLES AVE, 004XX W 42ND PL, 025XX N K...   

                                                IUCR  \
0  [0840, 2825, 1752, 0841, 0266, 5007, 0890, 175...   

                                        Primary Type  \
0  [THEFT, OTHER OFFENSE, OFFENSE INVOLVING CHILD...   

                                         Description  \
0  [FINANCIAL ID THEFT: OVER $300, HARASSMENT BY ...   

                                Location Description         Arrest  \
0  [RESIDENCE, OTHER, APARTMENT, RESIDENCE PORCH/...  [False, True]   

        Domestic                                               Beat  \
0  [False, True]  [424, 935, 1413, 2521, 2233, 1011, 531, 2222, ...   

                                            District  \
0  [4.0, 9.0, 14.0, 25.0, 22.0, 10.0, 5.0, 18.0, ...   

                                                Ward  \
0  [7.0, 11.0, 35.0, 31.0, 34.0, 24.0, 9.0, 21.0,...   

                                      Community Area  \
0  [46.0, 61.0, 22.0, 20.0, 49.0, 29.0, 50.0, 73....   

                                            FBI Code  \
0  [06, 26, 20, 02, 07, 17, 11, 10, 08B, 05, 15, ...   

                                        X Coordinate  \
0  [nan, 1173974.0, 1174948.0, 1182247.0, 1169911...   

                                        Y Coordinate  \
0  [nan, 1876757.0, 1831051.0, 1829375.0, 1844832...   

                                                Year  \
0  [2004.0, 2003.0, 2001.0, 2002.0, 41.789832136,...   

                                          Updated On  \
0  [08/17/2015 03:03:40 PM, 04/15/2016 08:55:02 A...   

                                            Latitude  \
0  [nan, 41.817229155999996, 41.691784636, 41.687...   

                                           Longitude  \
0  [nan, -87.637328162, -87.635115968, -87.608445...   

                                            Location  
0  [nan, (41.817229156, -87.637328162), (41.69178...

In [42]:
df_missingSummaries

Unnamed: 0 ID  Case Number Date Block IUCR Primary Type Description  \
df1          0  0            0    0     0    0            0           0   
df2          0  0            0    0     0    0            0           0   
df3          0  0  0.000223155    0     0    0            0           0   
df4          0  0  6.86477e-05    0     0    0            0           0   

    Location Description Arrest Domestic Beat     District         Ward  \
df1          0.000831811      0        0    0  0.000103976      36.3986   
df2           0.00133523      0        0    0  0.000267045  0.000801135   
df3             0.010823      0        0    0   0.00308698   0.00234313   
df4             0.113818      0        0    0  6.86477e-05  0.000961067   

    Community Area FBI Code X Coordinate Y Coordinate Year Updated On  \
df1        36.4046        0      1.59557      1.59557    0          0   
df2      0.0186397        0     0.488906     0.488906    0          0   
df3      0.0541152        0      1.06538      1.06538    0          0   
df4     0.00274591        0      2.54566      2.54566    0          0   

     Latitude Longitude  Location  
df1   1.59557   1.59562   1.59562  
df2  0.488906  0.488906  0.488906  
df3   1.06538   1.06538   1.06538  
df4   2.54566   2.54566   2.54566

In [61]:
def printUniqueSummaryValues (df): 
    for column in df.columns: 
        print(column + " has these many unique values: ")
        print(len(df[column][0]))

In [62]:
printUniqueSummaryValues(dftotal_unique)

Unnamed: 0 has these many unique values: 
6170812
ID has these many unique values: 
6170812
Case Number has these many unique values: 
6170473
Date has these many unique values: 
2451622
Block has these many unique values: 
58776
IUCR has these many unique values: 
398
Primary Type has these many unique values: 
35
Description has these many unique values: 
376
Location Description has these many unique values: 
173
Arrest has these many unique values: 
2
Domestic has these many unique values: 
2
Beat has these many unique values: 
304
District has these many unique values: 
27
Ward has these many unique values: 
51
Community Area has these many unique values: 
79
FBI Code has these many unique values: 
26
X Coordinate has these many unique values: 
78276
Y Coordinate has these many unique values: 
152136
Year has these many unique values: 
18
Updated On has these many unique values: 
1310
Latitude has these many unique values: 
864639
Longitude has these many unique values: 
838433
Lo

### Cleaning/Dropping columns 

We will drop the following information In the initial step: 
* Unnamed because it ID includes that information
* Ward because the dataset had a lot of missing data for df1
* Longitude & Latitude bc location includes both coordinates 
* Updated On b/c we are not interested in the data 
* X Coordinate & Y Coordinate, b/c we will retain location information anyway 
* Case Number in this case because we already have an ID method 


In [45]:
dfTotal.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [46]:
delInitialCols = ['Unnamed: 0', 'Case Number', 'Longitude', 'Latitude', 'Ward', 'X Coordinate', 'Y Coordinate', 'Updated On']

In [47]:
dfTotal = dfTotal.drop(delInitialCols, axis = 1) #actual deletion 

In [48]:
dfTotal.head()

ID                    Date                Block  IUCR  \
0  4786321  01/01/2004 12:01:00 AM    082XX S COLES AVE  0840   
1  4676906  03/01/2003 12:00:00 AM      004XX W 42ND PL  2825   
2  4789749  06/20/2004 11:00:00 AM  025XX N KIMBALL AVE  1752   
3  4789765  12/30/2004 08:00:00 PM   045XX W MONTANA ST  0840   
4  4677901  05/01/2003 01:00:00 AM   111XX S NORMAL AVE  0841   

                 Primary Type                     Description  \
0                       THEFT   FINANCIAL ID THEFT: OVER $300   
1               OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2  OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3                       THEFT   FINANCIAL ID THEFT: OVER $300   
4                       THEFT  FINANCIAL ID THEFT:$300 &UNDER   

  Location Description  Arrest  Domestic  Beat  District  Community Area  \
0            RESIDENCE   False     False   424       4.0            46.0   
1            RESIDENCE   False      True   935       9.0            61.0   
2            RESIDENCE   False     False  1413      14.0            22.0   
3                OTHER   False     False  2521      25.0            20.0   
4            RESIDENCE   False     False  2233      22.0            49.0   

  FBI Code    Year                       Location  
0       06  2004.0                            NaN  
1       26  2003.0  (41.817229156, -87.637328162)  
2       20  2004.0                            NaN  
3       06  2004.0                            NaN  
4       06  2003.0  (41.691784636, -87.635115968)

We will also delete:
* Community Area bc we will retain either beat or district 
* Block because it is too finegraine details that we do not need 

In [53]:
delInitialCols2 = ['Community Area', 'Block']

In [54]:
dfTotal = dfTotal.drop(delInitialCols2, axis = 1) #actual deletion 

In [55]:
dfTotal.head()

ID                    Date  IUCR                Primary Type  \
0  4786321  01/01/2004 12:01:00 AM  0840                       THEFT   
1  4676906  03/01/2003 12:00:00 AM  2825               OTHER OFFENSE   
2  4789749  06/20/2004 11:00:00 AM  1752  OFFENSE INVOLVING CHILDREN   
3  4789765  12/30/2004 08:00:00 PM  0840                       THEFT   
4  4677901  05/01/2003 01:00:00 AM  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   Beat  District FBI Code    Year                       Location  
0   424       4.0       06  2004.0                            NaN  
1   935       9.0       26  2003.0  (41.817229156, -87.637328162)  
2  1413      14.0       20  2004.0                            NaN  
3  2521      25.0       06  2004.0                            NaN  
4  2233      22.0       06  2003.0  (41.691784636, -87.635115968)

In [63]:
def printUniqueValues (df): 
    for column in df.columns: 
        print(column + " has these many unique values: ")
        print(len(df[column].unique()))

In [64]:
printUniqueValues(dfTotal)

ID has these many unique values: 
6170812
Date has these many unique values: 
2451622
IUCR has these many unique values: 
398
Primary Type has these many unique values: 
35
Description has these many unique values: 
376
Location Description has these many unique values: 
173
Arrest has these many unique values: 
2
Domestic has these many unique values: 
2
Beat has these many unique values: 
304
District has these many unique values: 
27
FBI Code has these many unique values: 
26
Year has these many unique values: 
18
Location has these many unique values: 
840086


We will drop Beat -- while this info has 0 missing data values, it has 304 different categories while District has 27 and has missing categories. This decision is made with the belief that we do not want to make our clusterings into 304 different ones but in a more general sense than that 

In [65]:
dfTotal = dfTotal.drop(['Beat'], axis = 1) #actual deletion 

In [67]:
len(dfTotal)

7941282

In [69]:
dfTotal = dfTotal.reset_index() #to realign the indices from the concatenation of four dataframesfrom before

In [70]:
dfTotal.head()

index       ID                    Date  IUCR                Primary Type  \
0      0  4786321  01/01/2004 12:01:00 AM  0840                       THEFT   
1      1  4676906  03/01/2003 12:00:00 AM  2825               OTHER OFFENSE   
2      2  4789749  06/20/2004 11:00:00 AM  1752  OFFENSE INVOLVING CHILDREN   
3      3  4789765  12/30/2004 08:00:00 PM  0840                       THEFT   
4      4  4677901  05/01/2003 01:00:00 AM  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District FBI Code    Year                       Location  
0       4.0       06  2004.0                            NaN  
1       9.0       26  2003.0  (41.817229156, -87.637328162)  
2      14.0       20  2004.0                            NaN  
3      25.0       06  2004.0                            NaN  
4      22.0       06  2003.0  (41.691784636, -87.635115968)

dfTotal.to_csv("../../crimesInChicagoData/dfTotal.csv") #exporting to a csv so we can use this dataset and have fewer crashes 

### Using the newly exported Total DF Csv Data set 

In [291]:
newTotalDf = pd.read_csv("../../crimesInChicagoData/dfTotal.csv")

In [292]:
newTotalDf = newTotalDf.drop(['Unnamed: 0'], axis = 1)

### Converting to Timeseries data of Date

In [293]:
changeToDateTime(newTotalDf, 'Date', dateFormat)

In [294]:
newTotalDf.head()

index       ID                Date  IUCR                Primary Type  \
0      0  4786321 2004-01-01 00:01:00  0840                       THEFT   
1      1  4676906 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2      2  4789749 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3      3  4789765 2004-12-30 20:00:00  0840                       THEFT   
4      4  4677901 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District FBI Code    Year                       Location  
0       4.0       06  2004.0                            NaN  
1       9.0       26  2003.0  (41.817229156, -87.637328162)  
2      14.0       20  2004.0                            NaN  
3      25.0       06  2004.0                            NaN  
4      22.0       06  2003.0  (41.691784636, -87.635115968)

### Removing Incorrect Data 
* there was 2 rows with the same information 
* found an error that indicated that there was a wrong data format -- so we just found the equivalent and deleted the ID

In [295]:
newTotalDf.loc[newTotalDf['Year'] == 41.789832136]

index       ID                Date  IUCR     Primary Type  \
1602847  1602847  1423259 2001-03-10 23:30:00  1340  CRIMINAL DAMAGE   

               Description     Location Description  Arrest  Domestic  \
1602847  TO STATE SUP PROP  CHA PARKING LOT/GROUNDS    True     False   

         District FBI Code       Year Location  
1602847       2.0       14  41.789832      NaN

In [296]:
newTotalDf.loc[newTotalDf['ID'] == 1423259]

index       ID                Date  IUCR     Primary Type  \
89258      89258  1423259 2001-03-10 23:30:00  1340  CRIMINAL DAMAGE   
1602847  1602847  1423259 2001-03-10 23:30:00  1340  CRIMINAL DAMAGE   

               Description     Location Description  Arrest  Domestic  \
89258    TO STATE SUP PROP  CHA PARKING LOT/GROUNDS    True     False   
1602847  TO STATE SUP PROP  CHA PARKING LOT/GROUNDS    True     False   

         District FBI Code         Year                       Location  
89258         2.0       14  2001.000000  (41.830443262, -87.628848503)  
1602847       2.0       14    41.789832                            NaN

In [297]:
newTotalDf.drop(newTotalDf.index[1602847], inplace=True)
newTotalDf.loc[newTotalDf['ID'] == 1423259]

index       ID                Date  IUCR     Primary Type  \
89258  89258  1423259 2001-03-10 23:30:00  1340  CRIMINAL DAMAGE   

             Description     Location Description  Arrest  Domestic  District  \
89258  TO STATE SUP PROP  CHA PARKING LOT/GROUNDS    True     False       2.0   

      FBI Code    Year                       Location  
89258       14  2001.0  (41.830443262, -87.628848503)

In [298]:
newTotalDf = newTotalDf.reset_index() #reset data 

In [299]:
newTotalDf.head()

level_0  index       ID                Date  IUCR  \
0        0      0  4786321 2004-01-01 00:01:00  0840   
1        1      1  4676906 2003-03-01 00:00:00  2825   
2        2      2  4789749 2004-06-20 11:00:00  1752   
3        3      3  4789765 2004-12-30 20:00:00  0840   
4        4      4  4677901 2003-05-01 01:00:00  0841   

                 Primary Type                     Description  \
0                       THEFT   FINANCIAL ID THEFT: OVER $300   
1               OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2  OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3                       THEFT   FINANCIAL ID THEFT: OVER $300   
4                       THEFT  FINANCIAL ID THEFT:$300 &UNDER   

  Location Description  Arrest  Domestic  District FBI Code    Year  \
0            RESIDENCE   False     False       4.0       06  2004.0   
1            RESIDENCE   False      True       9.0       26  2003.0   
2            RESIDENCE   False     False      14.0       20  2004.0   
3                OTHER   False     False      25.0       06  2004.0   
4            RESIDENCE   False     False      22.0       06  2003.0   

                        Location  
0                            NaN  
1  (41.817229156, -87.637328162)  
2                            NaN  
3                            NaN  
4  (41.691784636, -87.635115968)

### Unnest Location Data 

In [300]:
locationDf = newTotalDf['Location']
latitudeList = []
longitudeList = []
for location in locationDf:
    if(type(location) != str): #if it is nan 
        latitudeList.append(location)
        longitudeList.append(location)
    elif(type(location) == str): 
        coordinate = location.replace("(", "")
        coordinate = coordinate.replace(")", "")
        splitCoords = coordinate.split(",")
        latCoord = float(splitCoords[0])
        longCoord = float(splitCoords[1])
        latitudeList.append(latCoord)
        longitudeList.append(longCoord)
        
    else: 
        print(location)

In [301]:
latitudeList

[nan,
 41.817229156,
 nan,
 nan,
 41.691784636,
 nan,
 41.687020002,
 41.729712374,
 nan,
 41.869772159,
 41.724330486,
 nan,
 41.783163561,
 nan,
 41.831494381,
 41.778921304,
 41.911824751,
 41.791518681,
 41.834522994,
 nan,
 41.778539988,
 nan,
 41.771736514,
 41.745267715,
 41.975821967,
 41.896808547,
 41.766012869,
 nan,
 nan,
 nan,
 nan,
 41.87799981,
 41.921794873,
 41.772186514,
 41.843787854,
 41.789249657,
 41.931561031,
 41.88171118,
 nan,
 41.966409353,
 41.753507587,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 41.779814746,
 nan,
 nan,
 41.748448075,
 nan,
 41.852159338,
 41.8002858,
 41.921325664,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 41.968654176,
 41.657827784,
 41.902041291,
 41.905359186,
 41.782951712,
 nan,
 nan,
 41.830907597,
 41.964488675,
 41.774967846,
 nan,
 nan,
 nan,
 41.785980536,
 41.884673766,
 nan,
 nan,
 41.72913814,
 nan,
 41.72256088,
 41.757374833,
 41.924495568,
 41.727730234,
 41.947941831,
 nan,
 nan,
 nan,
 41.916908245,
 nan,
 nan,
 nan,
 nan,
 

In [302]:
latitudeDf = pd.DataFrame({'Latitude':latitudeList})
latitudeDf.head(10)

Latitude
0        NaN
1  41.817229
2        NaN
3        NaN
4  41.691785
5        NaN
6  41.687020
7  41.729712
8        NaN
9  41.869772

In [303]:
longitudeDf = pd.DataFrame({'Longitude':longitudeList})
longitudeDf.head(10)

Longitude
0        NaN
1 -87.637328
2        NaN
3        NaN
4 -87.635116
5        NaN
6 -87.608445
7 -87.653159
8        NaN
9 -87.708180

In [304]:
locationsDf = latitudeDf.join(longitudeDf)

In [305]:
locationsDf.head(20)

Latitude  Longitude
0         NaN        NaN
1   41.817229 -87.637328
2         NaN        NaN
3         NaN        NaN
4   41.691785 -87.635116
5         NaN        NaN
6   41.687020 -87.608445
7   41.729712 -87.653159
8         NaN        NaN
9   41.869772 -87.708180
10  41.724330 -87.638434
11        NaN        NaN
12  41.783164 -87.727773
13        NaN        NaN
14  41.831494 -87.667977
15  41.778921 -87.706365
16  41.911825 -87.715753
17  41.791519 -87.729099
18  41.834523 -87.682642
19        NaN        NaN

In [306]:
print(len(latitudeDf))
print(len(longitudeDf))
print(len(newTotalDf))
print(len(locationsDf))

7941281
7941281
7941281
7941281


In [307]:
newTotalDf =newTotalDf.join(locationsDf)

In [309]:
newTotalDf.head(20)

level_0  index       ID                Date  IUCR  \
0         0      0  4786321 2004-01-01 00:01:00  0840   
1         1      1  4676906 2003-03-01 00:00:00  2825   
2         2      2  4789749 2004-06-20 11:00:00  1752   
3         3      3  4789765 2004-12-30 20:00:00  0840   
4         4      4  4677901 2003-05-01 01:00:00  0841   
5         5      5  4838048 2004-08-01 00:01:00  0841   
6         6      6  4791194 2001-01-01 11:00:00  0266   
7         7      7  4679521 2003-03-15 00:00:00  5007   
8         8      8  4792195 2004-09-16 10:00:00  0890   
9         9      9  4680124 2003-01-01 00:00:00  0840   
10       10     10  4682244 2003-06-05 00:00:00  0840   
11       11     11  4795453 2004-11-01 00:01:00  0840   
12       12     12  4684808 2003-01-01 00:00:00  0840   
13       13     13  4800321 2004-11-01 00:00:00  0840   
14       14     14  4686904 2003-11-01 08:00:00  0840   
15       15     15  4687321 2003-11-15 09:00:00  0840   
16       16     16  4688004 2003-01-02 08:00:00  0840   
17       17     17  4803028 2001-01-01 00:00:00  1753   
18       18     18  4803606 2001-01-01 00:00:00  0840   
19       19     19  4808626 2002-06-14 16:00:00  0266   

                  Primary Type                     Description  \
0                        THEFT   FINANCIAL ID THEFT: OVER $300   
1                OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2   OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3                        THEFT   FINANCIAL ID THEFT: OVER $300   
4                        THEFT  FINANCIAL ID THEFT:$300 &UNDER   
5                        THEFT  FINANCIAL ID THEFT:$300 &UNDER   
6          CRIM SEXUAL ASSAULT                       PREDATORY   
7                OTHER OFFENSE         OTHER WEAPONS VIOLATION   
8                        THEFT                   FROM BUILDING   
9                        THEFT   FINANCIAL ID THEFT: OVER $300   
10                       THEFT   FINANCIAL ID THEFT: OVER $300   
11                       THEFT   FINANCIAL ID THEFT: OVER $300   
12                       THEFT   FINANCIAL ID THEFT: OVER $300   
13                       THEFT   FINANCIAL ID THEFT: OVER $300   
14                       THEFT   FINANCIAL ID THEFT: OVER $300   
15                       THEFT   FINANCIAL ID THEFT: OVER $300   
16                       THEFT   FINANCIAL ID THEFT: OVER $300   
17  OFFENSE INVOLVING CHILDREN   SEX ASSLT OF CHILD BY FAM MBR   
18                       THEFT   FINANCIAL ID THEFT: OVER $300   
19         CRIM SEXUAL ASSAULT                       PREDATORY   

       Location Description  Arrest  Domestic  District FBI Code    Year  \
0                 RESIDENCE   False     False       4.0       06  2004.0   
1                 RESIDENCE   False      True       9.0       26  2003.0   
2                 RESIDENCE   False     False      14.0       20  2004.0   
3                     OTHER   False     False      25.0       06  2004.0   
4                 RESIDENCE   False     False      22.0       06  2003.0   
5                 APARTMENT   False     False      10.0       06  2004.0   
6                 RESIDENCE    True      True       5.0       02  2001.0   
7   RESIDENCE PORCH/HALLWAY   False     False      22.0       26  2003.0   
8                 RESIDENCE   False     False      18.0       06  2004.0   
9                 RESIDENCE   False     False      11.0       06  2003.0   
10                RESIDENCE   False     False      22.0       06  2003.0   
11                RESIDENCE   False     False      20.0       06  2004.0   
12                RESIDENCE   False     False       8.0       06  2003.0   
13                RESIDENCE   False     False       4.0       06  2004.0   
14                RESIDENCE   False     False       9.0       06  2003.0   
15                APARTMENT   False     False       8.0       06  2003.0   
16                RESIDENCE   False     False      14.0       06  2003.0   
17                RESIDENCE    True     False      

In [282]:
newTotalDf= newTotalDf.drop(['index', 'Location'],axis =1)

In [326]:
newTotalDf.head(10)

level_0       ID                Date  IUCR                Primary Type  \
0        0  4786321 2004-01-01 00:01:00  0840                       THEFT   
1        1  4676906 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2        2  4789749 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3        3  4789765 2004-12-30 20:00:00  0840                       THEFT   
4        4  4677901 2003-05-01 01:00:00  0841                       THEFT   
5        5  4838048 2004-08-01 00:01:00  0841                       THEFT   
6        6  4791194 2001-01-01 11:00:00  0266         CRIM SEXUAL ASSAULT   
7        7  4679521 2003-03-15 00:00:00  5007               OTHER OFFENSE   
8        8  4792195 2004-09-16 10:00:00  0890                       THEFT   
9        9  4680124 2003-01-01 00:00:00  0840                       THEFT   

                      Description     Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300                RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE                RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER                RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                    OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER                RESIDENCE   False     False   
5  FINANCIAL ID THEFT:$300 &UNDER                APARTMENT   False     False   
6                       PREDATORY                RESIDENCE    True      True   
7         OTHER WEAPONS VIOLATION  RESIDENCE PORCH/HALLWAY   False     False   
8                   FROM BUILDING                RESIDENCE   False     False   
9   FINANCIAL ID THEFT: OVER $300                RESIDENCE   False     False   

   District FBI Code    Year   Latitude  Longitude  
0       4.0       06  2004.0        NaN        NaN  
1       9.0       26  2003.0  41.817229 -87.637328  
2      14.0       20  2004.0        NaN        NaN  
3      25.0       06  2004.0        NaN        NaN  
4      22.0       06  2003.0  41.691785 -87.635116  
5      10.0       06  2004.0        NaN        NaN  
6       5.0       02  2001.0  41.687020 -87.608445  
7      22.0       26  2003.0  41.729712 -87.653159  
8      18.0       06  2004.0        NaN        NaN  
9      11.0       06  2003.0  41.869772 -87.708180

In [327]:
newTotalDf.to_csv("../../crimesInChicagoData/newTotalDf.csv")

### Import the New Cleaned Dataset With Location Information
* Also note that because we put things in a float format some of the decimals rounded up, so we did lose information, but we don't need to look at information on geography with such detail 

In [2]:
import numpy as np  
import pandas as pd 
import math

In [3]:
dataset = pd.read_csv("../../crimesInChicagoData/newTotalDf.csv")

In [ ]:
dataset = dataset.drop(['level_0'], axis =1)

In [6]:
dataset.head()

Unnamed: 0       ID                 Date  IUCR                Primary Type  \
0           0  4786321  2004-01-01 00:01:00  0840                       THEFT   
1           1  4676906  2003-03-01 00:00:00  2825               OTHER OFFENSE   
2           2  4789749  2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3           3  4789765  2004-12-30 20:00:00  0840                       THEFT   
4           4  4677901  2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District FBI Code    Year   Latitude  Longitude  
0       4.0       06  2004.0        NaN        NaN  
1       9.0       26  2003.0  41.817229 -87.637328  
2      14.0       20  2004.0        NaN        NaN  
3      25.0       06  2004.0        NaN        NaN  
4      22.0       06  2003.0  41.691785 -87.635116

### Using the District Data to fill in the Missing Data Value of Location 
#### Part I: Filling in NaN Data of lognitude and latitude of districts that do not have a missing district designation
* District corresponds to a specific region.
* In this section we subdivide the data by district then find the "average" location for each district 
* That average will be used to fill in the location missing value 

In [11]:
uniqueDistricts = list(dataset['District'].unique())

uniqueDistricts.remove(uniqueDistricts[24]) #removed the nan 

In [12]:
uniqueDistricts

[4.0,
 9.0,
 14.0,
 25.0,
 22.0,
 10.0,
 5.0,
 18.0,
 11.0,
 20.0,
 8.0,
 7.0,
 1.0,
 16.0,
 15.0,
 3.0,
 6.0,
 2.0,
 19.0,
 12.0,
 24.0,
 17.0,
 31.0,
 21.0,
 23.0,
 13.0]

In [13]:
def subsetLists(df, subsetCategories, columnName): 
    uniqueSubsetsList = []
    for category in subsetCategories: 
        dfSubset = df.loc[df[columnName] == category] 
        uniqueSubsetsList.append(dfSubset)
    return uniqueSubsetsList

In [14]:
uniqueDistrictsList = subsetLists(dataset, uniqueDistricts, 'District')

In [28]:
for i in range(len(uniqueDistrictsList)):
    print(str(uniqueDistrictsList[i]['District'].unique()) + " district is consisted of the follwing # of districts: ")
    print(len(uniqueDistrictsList[i]))

[ 4.] district is consisted of the follwing # of districts: 
453894
[ 9.] district is consisted of the follwing # of districts: 
397942
[ 14.] district is consisted of the follwing # of districts: 
314642
[ 25.] district is consisted of the follwing # of districts: 
463938
[ 22.] district is consisted of the follwing # of districts: 
263055
[ 10.] district is consisted of the follwing # of districts: 
335491
[ 5.] district is consisted of the follwing # of districts: 
354567
[ 18.] district is consisted of the follwing # of districts: 
337197
[ 11.] district is consisted of the follwing # of districts: 
498775
[ 20.] district is consisted of the follwing # of districts: 
137310
[ 8.] district is consisted of the follwing # of districts: 
550011
[ 7.] district is consisted of the follwing # of districts: 
476524
[ 1.] district is consisted of the follwing # of districts: 
291073
[ 16.] district is consisted of the follwing # of districts: 
263442
[ 15.] district is consisted of the foll

In [42]:
def fillMissingWithAverage(columnName, df):
    totalSum = df[columnName].sum()
    totalCount = df[columnName].describe()['count']
    average = totalSum/totalCount 
    df[columnName].fillna(average, inplace = True)

In [43]:
for dataframe in uniqueDistrictsList:
    fillMissingWithAverage('Latitude', dataframe)
    fillMissingWithAverage('Longitude', dataframe)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/generic.py:3660: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


#### Part II: Filling in Districts with Missing Values, but nonempty longitude/latitude data

In [90]:
avgCoordsList = []
for dataFrame in  uniqueDistrictsList:
    coords = []
    lat = dataFrame['Latitude'].sum()/dataFrame['Latitude'].describe()['count']
    long = dataFrame['Longitude'].sum()/dataFrame['Longitude'].describe()['count']
    coords.append(lat)
    coords.append(long)
    coords.append(str(dataFrame['District'].unique()[0]))
    avgCoordsList.append(coords)
    print(coords)

[41.734106465357918, -87.563621465897356, '4.0']
[41.814834324982158, -87.665279593948156, '9.0']
[41.915619987842391, -87.694019153700154, '14.0']
[41.919054267813586, -87.752178204964579, '25.0']
[41.708615973820699, -87.658508927398941, '22.0']
[41.853446584131227, -87.712625046552503, '10.0']
[41.687693669783748, -87.62293524998104, '5.0']
[41.902862496469353, -87.636098052120047, '18.0']
[41.882593721342076, -87.71915927439089, '11.0']
[41.978504465170332, -87.671868012214816, '20.0']
[41.77828299100242, -87.715424268281978, '8.0']
[41.775835445561057, -87.653798495789147, '7.0']
[41.871889066502099, -87.62891553227513, '1.0']
[41.964588567605986, -87.796818568950187, '16.0']
[41.886361521048535, -87.758102961052728, '15.0']
[41.770891499332414, -87.597182482763628, '3.0']
[41.745504921274303, -87.63270767909394, '6.0']
[41.810943162372638, -87.613066009139018, '2.0']
[41.947739249024593, -87.660401833179506, '19.0']
[41.880340460211137, -87.672350340257438, '12.0']
[42.0056650132

In [76]:
nullDistricts = dataset.loc[dataset['District'].isnull() ]

In [81]:
nullDistricts['Latitude'].isnull().sum()
nullDistricts['Longitude'].isnull().sum()

0

In [82]:
nullDistricts

Unnamed: 0       ID                 Date  IUCR         Primary Type  \
1802296     1802296  3575885  2004-06-26 10:00:00  1150   DECEPTIVE PRACTICE   
1817065     1817065  3596991  2004-10-14 15:41:00  0610             BURGLARY   
1969393     1969393  4740376  2006-05-10 19:40:00  0460              BATTERY   
2503109     2503109  4740376  2006-05-10 19:40:00  0460              BATTERY   
3242555     3242555  6405961  2007-11-16 00:01:00  1150   DECEPTIVE PRACTICE   
3242582     3242582  6420740  2007-01-07 05:00:00  1140   DECEPTIVE PRACTICE   
3636356     3636356  4437079  2005-11-10 10:10:00  2825        OTHER OFFENSE   
4031834     4031834  6376239  2008-07-18 20:00:00  1310      CRIMINAL DAMAGE   
4058377     4058377  6420058  2008-07-13 18:08:55  1811            NARCOTICS   
4077370     4077370  6451103  2008-08-23 21:23:00  1563          SEX OFFENSE   
4082462     4082462  6459556  2008-08-26 23:40:00  0320              ROBBERY   
4084439     4084439  6462999  2008-08-23 12:00:00  1122   DECEPTIVE PRACTICE   
4098625     4098625  6486045  2008-09-07 13:00:00  4387        OTHER OFFENSE   
4099704     4099704  6488159  2008-09-09 17:00:00  0910  MOTOR VEHICLE THEFT   
4115300     4115300  6514077  2008-09-21 22:00:00  0550              ASSAULT   
4116497     4116497  6516041  2008-09-22 21:10:00  0560              ASSAULT   
4118830     4118830  6519692  2008-09-20 20:00:00  1563          SEX OFFENSE   
4128198     4128198  6534030  2008-09-03 12:00:00  0910  MOTOR VEHICLE THEFT   
4128561     4128561  6534600  2008-10-03 12:50:00  1130   DECEPTIVE PRACTICE   
4138694     4138694  6549802  2008-10-11 15:00:00  0910  MOTOR VEHICLE THEFT   
4142812     4142812  6555854  2008-10-15 16:10:00  0610             BURGLARY   
4145692     4145692  6560428  2008-10-16 23:00:00  0313              ROBBERY   
4146443     4146443  6561442  2008-10-18 06:30:00  0291  CRIM SEXUAL ASSAULT   
4146902     4146902  6562039  2008-10-17 22:39:00  0320              ROBBERY   
4151616     4151616  6569778  2008-10-22 08:30:00  0610             BURGLARY   
4155881     4155881  6576124  2008-08-21 14:30:00  1152   DECEPTIVE PRACTICE   
4156878     4156878  6577661  2008-10-17 12:00:00  0843                THEFT   
4157284     4157284  6578354  2008-10-27 21:19:00  0920  MOTOR VEHICLE THEFT   
4157930     4157930  6579338  2008-10-25 17:45:00  0330              ROBBERY   
4158170     4158170  6579790  2008-10-28 12:30:00  0890                THEFT   
...             ...      ...                  ...   ...                  ...   
5290307     5290307  6555854  2008-10-15 16:10:00  0610             BURGLARY   
5293187     5293187  6560428  2008-10-16 23:00:00  0313              ROBBERY   
5293938     5293938  6561442  2008-10-18 06:30:00  0291  CRIM SEXUAL ASSAULT   
5294397     5294397  6562039  2008-10-17 22:39:00  0320              ROBBERY   
5299111     5299111  6569778  2008-10-22 08:30:00  0610             BURGLARY   
5303376     5303376  6576124  2008-08-21 14:30:00  1152   DECEPTIVE PRACTICE   
5304373     5304373  6577661  2008-10-17 12:00:00  0843                THEFT   
5304779     5304779  6578354  2008-10-27 21:19:00  0920  MOTOR VEHICLE THEFT   
5305425     5305425  6579338  2008-10-25 17:45:00  0330              ROBBERY   
5305665     5305665  6579790  2008-10-28 12:30:00  0890                THEFT   
5305758     5305758  6580042  2008-10-21 20:15:00  1506         PROSTITUTION   
5305938     5305938  6580306  2008-10-24 17:45:00  1345      CRIMINAL DAMAGE   
5305945     5305945  6580316  2008-10-27 19:14:20  0560              ASSAULT   
5306023     5306023  6580405  2008-10-28 19:10:00  2900    WEAPONS VIOLATION   
5306297     5306297  6580757  2008-10-26 03:02:00  0486              BATTERY   
5307406     5307406  6582700  2008-10-29 05:09:00  0430              BATTERY   
5307772     5307772  6583267  2008-10-14 21:44:00  143A    WEAPONS VIOLATION   
5308929     5308929  6585169  2008-10-10 23:57:00  0554              ASSAULT   
5310725  

##### Reverse Engineering :
* Using the average latitude, longitude data to predict the nullDistricts Designation
* Using the euclidean distance formula: sqrt((x2-x1)^2+ (y2-y1)^2) 

In [86]:
def euclideanDistance(x1, x2, y1, y2):
    return math.sqrt((x2 -x1)**2 + (y2-y1)**2) 

In [117]:
for observation in nullDistricts.index:
    origLat = nullDistricts['Latitude'][observation]
    origLong = nullDistricts['Longitude'][observation]
    distanceList = []
    for coord in avgCoordsList: 
        avgLat = coord[0]
        avgLong = coord[1]
        distanceList.append(euclideanDistance(origLat, avgLat, origLong, avgLong))
    
    shortestDistance = min(distanceList)
    shortestDistanceIndex = distanceList.index(shortestDistance)
    district = float(avgCoordsList[shortestDistanceIndex][2])
    
    nullDistricts['District'][observation] = district
#     print("this is the distance list: ")
#     print(distanceList)
#     print("\n")
#     print("this is the shortest distance:" )
#     print(shortestDistance)
#     print("\n")
#     print("this is the index: ")
#     print(shortestDistanceIndex)
#     print("\n")  
#     print("this is the district:")
#     print(district)
#     print("******************************\n")

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2910: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [119]:
nullDistricts['District'].isnull().sum() #check if districts have been successfully filled

0

In [124]:
allDistricts = uniqueDistrictsList 

allDistricts.append(nullDistricts)

### Reseting datset to the new datset with non missing Longitude, Latitude & District Values

In [131]:
dataset = pd.concat(allDistricts, axis = 0)

In [135]:
dataset.sort_index()

Unnamed: 0        ID                 Date  IUCR  \
0                 0   4786321  2004-01-01 00:01:00  0840   
1                 1   4676906  2003-03-01 00:00:00  2825   
2                 2   4789749  2004-06-20 11:00:00  1752   
3                 3   4789765  2004-12-30 20:00:00  0840   
4                 4   4677901  2003-05-01 01:00:00  0841   
5                 5   4838048  2004-08-01 00:01:00  0841   
6                 6   4791194  2001-01-01 11:00:00  0266   
7                 7   4679521  2003-03-15 00:00:00  5007   
8                 8   4792195  2004-09-16 10:00:00  0890   
9                 9   4680124  2003-01-01 00:00:00  0840   
10               10   4682244  2003-06-05 00:00:00  0840   
11               11   4795453  2004-11-01 00:01:00  0840   
12               12   4684808  2003-01-01 00:00:00  0840   
13               13   4800321  2004-11-01 00:00:00  0840   
14               14   4686904  2003-11-01 08:00:00  0840   
15               15   4687321  2003-11-15 09:00:00  0840   
16               16   4688004  2003-01-02 08:00:00  0840   
17               17   4803028  2001-01-01 00:00:00  1753   
18               18   4803606  2001-01-01 00:00:00  0840   
19               19   4808626  2002-06-14 16:00:00  0266   
20               20   4809221  2001-06-16 08:00:00  1754   
21               21   4809407  2004-01-01 00:00:00  0840   
22               22   4695836  2003-10-31 14:00:00  0910   
23               23   4696148  2003-01-01 09:00:00  0840   
24               24   4696401  2003-01-01 00:00:00  0810   
25               25   4811727  2001-06-01 08:00:00  1563   
26               26   4699493  2003-01-01 12:00:00  0840   
27               27   4815217  2004-12-13 13:00:00  1130   
28               28   4815972  2004-10-06 09:00:00  0841   
29               29   4817129  2004-01-12 15:00:00  2825   
...             ...       ...                  ...   ...   
7941251     7941251  10508639  2016-05-03 22:32:00  051A   
7941252     7941252  10508640  2016-05-03 22:07:00  1811   
7941253     7941253  10508641  2016-05-03 22:31:00  143A   
7941254     7941254  10508642  2016-05-03 22:45:00  051A   
7941255     7941255  10508643  2016-05-03 21:00:00  0560   
7941256     7941256  10508644  2016-05-03 19:13:00  0460   
7941257     7941257  10508646  2016-05-03 19:45:00  0460   
7941258     7941258  10508647  2016-05-03 20:56:00  0460   
7941259     7941259  10508648  2016-05-03 22:10:00  0520   
7941260     7941260  10508649  2016-05-03 22:15:00  4387   
7941261     7941261  10508650  2016-05-03 17:00:00  031A   
7941262     7941262  10508653  2016-05-03 23:58:00  0520   
7941263     7941263  10508656  2016-05-03 15:15:00  1720   
7941264     7941264  10508658  2016-05-03 23:50:00  4625   
7941265     7941265  10508659  2016-05-03 23:38:00  0460   
7941266     7941266  10508661  2016-05-03 20:44:00  041A   
7941267     7941267  10508662  2016-05-03 08:00:00  5001   
7941268     7941268  10508663  2016-05-03 22:10:00  1563   
7941269     7941269  10508664  2016-05-03 23:35:00  0560   
7941270     7941270  10508665  2016-05-03 22:15:00  1310   
7941271     7941271  10508666  2016-05-03 23:30:00  0320   
7941272     7941272  10508671  2016-05-03 23:50:00  502P   
7941273     7941273  10508672  2016-05-03 22:25:00  0460   
7941274     7941274  10508675  2016-05-03 23:00:00  0320   
7941275     7941275  10508678  2016-05-03 23:28:00  041A   
7941276     7941276  10508679  2016-05-03 23:33:00  0486   
7941277     7941277  10508680  2016-05-03 23:30:00  1310   
7941278     7941278  10508681  2016-05-03 00:15:00  041A   
7941279     7941279  10508690  2016-05-03 21:07:00  0486   
7941280     7941280  10508692  2016-05-03 23:38:00  5007   

                       Primary Type                        Description  \
0                             THEFT      FINANCIAL ID THEFT: OVER $300   
1                     OTHER OFFENSE            HARASSMENT BY TELEPHONE   
2        OFFENSE INVOLVING CHILDREN      AGG CRIM SEX ABUSE FAM ME

In [136]:
dataset.describe()

Unnamed: 0            ID      District          Year      Latitude  \
count  7.941281e+06  7.941281e+06  7.941281e+06  7.941281e+06  7.941281e+06   
mean   3.970640e+06  5.926071e+06  1.131216e+01  2.007672e+03  4.184150e+01   
std    2.292451e+06  2.568290e+06  6.944530e+00  4.064012e+00  9.135010e-02   
min    0.000000e+00  6.340000e+02  1.000000e+00  2.001000e+03  3.661945e+01   
25%    1.985320e+06  3.853210e+06  6.000000e+00  2.005000e+03  4.176862e+01   
50%    3.970640e+06  6.165079e+06  1.000000e+01  2.008000e+03  4.185387e+01   
75%    5.955960e+06  7.716590e+06  1.700000e+01  2.010000e+03  4.190697e+01   
max    7.941280e+06  1.082788e+07  3.100000e+01  2.017000e+03  4.202291e+01   

          Longitude  
count  7.941281e+06  
mean  -8.767201e+01  
std    6.319202e-02  
min   -9.168657e+01  
25%   -8.771409e+01  
50%   -8.766638e+01  
75%   -8.762858e+01  
max   -8.752453e+01

In [139]:
dataset.isnull().sum()

Unnamed: 0                 0
ID                         0
Date                       0
IUCR                       0
Primary Type               0
Description                0
Location Description    1990
Arrest                     0
Domestic                   0
District                   0
FBI Code                   0
Year                       0
Latitude                   0
Longitude                  0
dtype: int64

### Filling in Missing Location Description Values 
* The summary tells us that there are 1990 observations of missing location values 

In [140]:
locDescriptionDf = dataset.loc[dataset['Location Description'].isnull()] 

In [142]:
locDescriptionDf

Unnamed: 0        ID                 Date  IUCR         Primary Type  \
518707       518707   1959008  2002-01-26 16:25:00  2820        OTHER OFFENSE   
770238       770238   2279950  2002-08-07 04:00:00  0261  CRIM SEXUAL ASSAULT   
1923399     1923399  10645212  2004-08-17 10:05:00  1153   DECEPTIVE PRACTICE   
3797266     3797266  10530814  2011-12-01 11:00:00  1153   DECEPTIVE PRACTICE   
3809308     3809308   8281441  2011-09-25 08:50:00  0890                THEFT   
3810892     3810892   9862543  2011-11-18 18:00:00  1153   DECEPTIVE PRACTICE   
4956803     4956803   8281441  2011-09-25 08:50:00  0890                THEFT   
4958387     4958387   9862543  2011-11-18 18:00:00  1153   DECEPTIVE PRACTICE   
6254420     6254420   8076790  2011-05-19 21:00:00  0820                THEFT   
6483015     6483015  10428831  2011-04-01 21:00:00  1153   DECEPTIVE PRACTICE   
6484255     6484255  10648347  2011-03-21 21:31:00  1153   DECEPTIVE PRACTICE   
6486497     6486497  10512796  2016-05-06 09:45:00  1153   DECEPTIVE PRACTICE   
6490386     6490386  10520073  2016-05-10 10:45:00  1154   DECEPTIVE PRACTICE   
6512937     6512937   9887465  2013-01-10 02:45:00  1153   DECEPTIVE PRACTICE   
6561460     6561460  10236194  2014-10-06 16:47:00  1154   DECEPTIVE PRACTICE   
6586549     6586549   8465780  2012-02-01 06:25:00  0820                THEFT   
6752428     6752428   8728358  2012-07-27 16:40:00  0820                THEFT   
6798674     6798674   8802908  2012-09-13 07:25:00  0820                THEFT   
7198267     7198267   9449747  2013-12-13 10:20:00  0820                THEFT   
7402420     7402420   9801254  2014-06-01 10:40:00  1153   DECEPTIVE PRACTICE   
7412588     7412588   9818578  2014-10-14 13:40:00  1153   DECEPTIVE PRACTICE   
7562797     7562797  10163757  2015-07-20 19:00:00  1154   DECEPTIVE PRACTICE   
7582466     7582466  10198252  2015-08-03 10:15:00  1153   DECEPTIVE PRACTICE   
7651915     7651915  10318995  2015-10-26 11:10:00  1153   DECEPTIVE PRACTICE   
7689513     7689513  10382707  2015-12-22 17:05:00  1153   DECEPTIVE PRACTICE   
7695516     7695516  10449350  2016-03-04 14:30:00  1150   DECEPTIVE PRACTICE   
7699004     7699004  10389202  2016-01-16 14:00:00  0610             BURGLARY   
7713787     7713787  10428873  2016-02-26 10:10:00  1153   DECEPTIVE PRACTICE   
7750737     7750737  10331869  2015-07-13 00:05:00  1153   DECEPTIVE PRACTICE   
7751779     7751779  10394643  2015-12-22 00:01:00  1153   DECEPTIVE PRACTICE   
...             ...       ...                  ...   ...                  ...   
7764531     7764531  10552670  2016-05-29 00:00:00  1153   DECEPTIVE PRACTICE   
7766193     7766193  10555521  2015-01-05 10:50:00  1153   DECEPTIVE PRACTICE   
7773024     7773024  10565895  2016-06-05 15:50:00  1153   DECEPTIVE PRACTICE   
7774442     7774442  10565861  2016-06-14 08:39:00  1153   DECEPTIVE PRACTICE   
7782709     7782709  10581203  2014-06-02 12:05:00  1153   DECEPTIVE PRACTICE   
7785593     7785593  10582505  2016-06-27 12:25:00  1154   DECEPTIVE PRACTICE   
7786278     7786278  10581204  2016-06-28 19:20:00  1154   DECEPTIVE PRACTICE   
7793921     7793921  10596491  2016-06-16 13:00:00  1153   DECEPTIVE PRACTICE   
7801970     7801970  10605399  2016-07-08 14:00:00  1153   DECEPTIVE PRACTICE   
7811785     7811785  10625982  2015-08-01 16:20:00  1154   DECEPTIVE PRACTICE   
7820105     7820105  10632700  2016-08-08 15:35:00  1154   DECEPTIVE PRACTICE   
7822451     7822451  10639150  2016-08-11 13:35:00  1153   DECEPTIVE PRACTICE   
7825266     7825266  10646671  2016-08-16 01:00:00  1153   DECEPTIVE PRACTICE   
7827295     7827295  10648357  2016-08-19 13:30:00  1153   DECEPTIVE PRACTICE   
7832020     7832020  10663198  2016-08-18 09:40:00  1153   DECEPTIVE PRACTICE   
7848965     7848965  10684975  2016-09-14 20:25:00  1153   DECEPTIVE PRACTICE   
7849865     7849865  10686941  2016-09-15 15:00:00  1153   DECEPTIVE PRACTICE   
7853450     7853450  10688316  2016-09-2

There are no unique identifiers that could help us what the data in 'Location Description' should be

In [148]:
dataset['Location Description'].unique()

array(['RESIDENCE', 'BANK', 'APARTMENT', 'STREET',
       'CHURCH/SYNAGOGUE/PLACE OF WORSHIP', 'COMMERCIAL / BUSINESS OFFICE',
       'SMALL RETAIL STORE', 'HOSPITAL BUILDING/GROUNDS', 'OTHER',
       'CURRENCY EXCHANGE', 'RESIDENTIAL YARD (FRONT/BACK)', 'RESTAURANT',
       'PORCH', 'VESTIBULE', 'YARD', 'AUTO', 'VACANT LOT', 'PARKING LOT',
       'PARKING LOT/GARAGE(NON.RESID.)', 'GROCERY FOOD STORE',
       'DEPARTMENT STORE', 'TAVERN/LIQUOR STORE',
       'SCHOOL, PUBLIC, BUILDING', 'DRUG STORE', 'RESIDENCE-GARAGE',
       'VEHICLE NON-COMMERCIAL', 'CTA BUS',
       'POLICE FACILITY/VEH PARKING LOT', 'CHA PARKING LOT/GROUNDS',
       'CHA APARTMENT', 'SCHOOL, PUBLIC, GROUNDS',
       'GOVERNMENT BUILDING/PROPERTY', 'SIDEWALK', 'VACANT LOT/LAND',
       'CONSTRUCTION SITE', 'FACTORY/MANUFACTURING BUILDING',
       'BOAT/WATERCRAFT', 'LAKEFRONT/WATERFRONT/RIVERBANK', 'ALLEY',
       'BASEMENT', 'GANGWAY', 'BOWLING ALLEY',
       'OTHER COMMERCIAL TRANSPORTATION', 'RESIDENCE PORCH/HALL

There is no identifier for 'unknown' locations so we will create one with UNKNOWN and fill missing values with it

In [149]:
dataset['Location Description'].fillna('UNKNOWN', inplace = True)

In [150]:
dataset.isnull().sum()

Unnamed: 0              0
ID                      0
Date                    0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
District                0
FBI Code                0
Year                    0
Latitude                0
Longitude               0
dtype: int64

### New Dataset with No Missing Values 

In [160]:
dataset = dataset.sort_index()

In [162]:
dataset.to_csv("../../crimesInChicagoData/dataset.csv")

In [163]:
cleanData = pd.read_csv("../../crimesInChicagoData/dataset.csv")

In [164]:
cleanData

Unnamed: 0  Unnamed: 0.1        ID                 Date  IUCR  \
0                 0             0   4786321  2004-01-01 00:01:00  0840   
1                 1             1   4676906  2003-03-01 00:00:00  2825   
2                 2             2   4789749  2004-06-20 11:00:00  1752   
3                 3             3   4789765  2004-12-30 20:00:00  0840   
4                 4             4   4677901  2003-05-01 01:00:00  0841   
5                 5             5   4838048  2004-08-01 00:01:00  0841   
6                 6             6   4791194  2001-01-01 11:00:00  0266   
7                 7             7   4679521  2003-03-15 00:00:00  5007   
8                 8             8   4792195  2004-09-16 10:00:00  0890   
9                 9             9   4680124  2003-01-01 00:00:00  0840   
10               10            10   4682244  2003-06-05 00:00:00  0840   
11               11            11   4795453  2004-11-01 00:01:00  0840   
12               12            12   4684808  2003-01-01 00:00:00  0840   
13               13            13   4800321  2004-11-01 00:00:00  0840   
14               14            14   4686904  2003-11-01 08:00:00  0840   
15               15            15   4687321  2003-11-15 09:00:00  0840   
16               16            16   4688004  2003-01-02 08:00:00  0840   
17               17            17   4803028  2001-01-01 00:00:00  1753   
18               18            18   4803606  2001-01-01 00:00:00  0840   
19               19            19   4808626  2002-06-14 16:00:00  0266   
20               20            20   4809221  2001-06-16 08:00:00  1754   
21               21            21   4809407  2004-01-01 00:00:00  0840   
22               22            22   4695836  2003-10-31 14:00:00  0910   
23               23            23   4696148  2003-01-01 09:00:00  0840   
24               24            24   4696401  2003-01-01 00:00:00  0810   
25               25            25   4811727  2001-06-01 08:00:00  1563   
26               26            26   4699493  2003-01-01 12:00:00  0840   
27               27            27   4815217  2004-12-13 13:00:00  1130   
28               28            28   4815972  2004-10-06 09:00:00  0841   
29               29            29   4817129  2004-01-12 15:00:00  2825   
...             ...           ...       ...                  ...   ...   
7941251     7941251       7941251  10508639  2016-05-03 22:32:00  051A   
7941252     7941252       7941252  10508640  2016-05-03 22:07:00  1811   
7941253     7941253       7941253  10508641  2016-05-03 22:31:00  143A   
7941254     7941254       7941254  10508642  2016-05-03 22:45:00  051A   
7941255     7941255       7941255  10508643  2016-05-03 21:00:00  0560   
7941256     7941256       7941256  10508644  2016-05-03 19:13:00  0460   
7941257     7941257       7941257  10508646  2016-05-03 19:45:00  0460   
7941258     7941258       7941258  10508647  2016-05-03 20:56:00  0460   
7941259     7941259       7941259  10508648  2016-05-03 22:10:00  0520   
7941260     7941260       7941260  10508649  2016-05-03 22:15:00  4387   
7941261     7941261       7941261  10508650  2016-05-03 17:00:00  031A   
7941262     7941262       7941262  10508653  2016-05-03 23:58:00  0520   
7941263     7941263       7941263  10508656  2016-05-03 15:15:00  1720   
7941264     7941264       7941264  10508658  2016-05-03 23:50:00  4625   
7941265     7941265       7941265  10508659  2016-05-03 23:38:00  0460   
7941266     7941266       7941266  10508661  2016-05-03 20:44:00  041A   
7941267     7941267       7941267  10508662  2016-05-03 08:00:00  5001   
7941268     7941268       7941268  10508663  2016-05-03 22:10:00  1563   
7941269     7941269       7941269  10508664  2016-05-03 23:35:00  0560   
7941270     7941270       7941270  10508665  2016-05-03 22:15:00  1310   
7941271     7941271       7941271  10508666  2016-05-03 23:30:00  0320   
7941272     7941272       7941272  10508671  2016-05-03 23:50:00  502P   
7941273     7